### ISSUE TITLE
Package Geo-replication

### ISSUE THREAD

**Issue by droyad** [OctopusDeploy/Specs#14]
At the moment, when deployments are done, packages are downloaded multiple times directly to the tentacles or uploaded from the Octopus server. This is a problem if the NuGet feed/Octopus server is remote (relatively slow network) in regards to those tentacles, as we transfer more data than needed.

**Comment by droyad** [comment]
We could solve this by creating a Package caching service. Each Tentacle/endpoint would specify which caching service is closest to them.

This could have the following capabilities:
1) Packages optionally auto replicate from the main server to any of these caches
1) On Deployment, the server pushes packages to the caches involved in the deployment
1) A tentacle can request a package from a cache, which then knows how to fetch it. Not sure if this is required though.

**Comment by slewis74** [comment]
3 is definitely important for some customers. Had a conversation with one yesterday where they were doing exactly that type of caching, with on-demand fetch. They have a custom solution at the moment.

**Comment by droyad** [comment]
What if we add a concept of `Network` into OD (perhaps linked to the ODCM concept). Each network could have a list of package caches in order of preference (providing redundancy). A tentacle would belong to a network.

**Comment by droyad** [comment]
@slewis74 Why would they need on-demand? The OD server should know which packages are involved and could do 2.

**Comment by zentron** [comment]
I think its more than just package caching. It seems like you might want some sort of "slimmed down" Octopus server task runner that can orchestrate the task running since the Tentacle->Server communications can be part of the perf issues (think of large logs etc). This would include package staging.

 This would work well with a hosted octopus scenario where we also might want to offload task orchestration from the main server.

**Comment by droyad** [comment]
@zentron The logs would need to be transferred back centrally in any case. At the moment they are only transferred once, so why not keep it as is? Is it because the network is being saturated by many tentacles reporting? Do loose the realtime feedback then?

**Comment by zentron** [comment]
@droyad agreed in principle, but do users want the task completed quick, with a slight increase delay in messaging, or slow with messaging quick? I think it comes down to latency as well as bandwidth with these distance-deployments though

**Comment by slewis74** [comment]
@droyad true, if OD server can optimize a single push to the cache when it knows it's about to deploy to the Tentacle(s) then it would probably cover their scenario.  Without the builtin support, their custom solution pulls the package when the first Tentacle requests it. 

Package sizes are 3GB+ so key is it needs to replicate, but not too eagerly.

**Comment by michaelnoonan** [comment]
This is one benefit you'd get by indicating a release is approved for a particular environment. One way you could do that in a Remote Release Promotions world would be to "Promote 2.5.1 to Prod Space". Octopus could interpret that as the command to replicate packages.

Another option would be if we introduced the concept of Release Approvals into Lifecycles - once a Release has been Approved, Octopus could kick off the replication to all the Environments in the Lifecycle up until the next Approval.

Very much just food for thought. But it does feel like we'd benefit from an "explicit user action" that indicates the packages should replicate.

**Comment by zentron** [comment]
@michaelnoonan this is part of what im thinking. It would be similar in some sense to a promotion, but in doing so it just stages the packages to the appropriate locations for when deployment takes place without the need for a whole bundle/receipt process

**Comment by slewis74** [comment]
I think in the scenario in question it was 1 Space and 1 Env, but groups of Tentacles in different countries.  So during a deployment to that Env, Tentacles should grab packages from the nearest cache rather than the OD server itself.

**Comment by droyad** [comment]
@michaelnoonan in promotions I agree. This feature would also be used within a space.

**Comment by droyad** [comment]
@zentron regarding logs, Instead of having an aggregator, why not build that right into the tentacle? It should return it's success and then send the log at a later time. Adding that aggregator adds an extra hop without benefit since each log is unique. 

Also it means that if the aggregator goes down, we lose HA, unless we cluster the aggregators themselves, doubling the amount of transfer from the tentacles and complexity.

**Comment by matt-richardson** [comment]
Something like JFrog's [Artifactory](https://www.jfrog.com/artifactory/) can do package replication, and is laser focused on packages and package management. AFAIK [Sonatype Nexus](https://www.sonatype.com/nexus-repository-sonatype) does replication too.

Would we not be better recommending something like that, instead of building our own?

**Comment by droyad** [comment]
@matt-richardson Possibly, and I think that's what people are doing. The scenario is a customer with 10 sites and 1TB of packages

We could use those product, but I think we still need to build something into the product. The main advantages I see are (not knowing those other products in detail):

- We know to which sites the packages should go to, so not all mirrors need to have every package or every version (also we can do JIT or just ahead of time distribution)
- We can apply retention policies based on releases
- We can do delta compression
- We can restrict those feeds so that packages from project A can't be seen by project B

If we did use a third party for part of the solution, we would need to restrict it to specific ones we support, since we need to push zips as well as nupkgs, which AFAIK no standard protocol exists.

**Comment by matt-richardson** [comment]
In a world with unlimited resources, I'd probably agree with you. However, with limited resources available to us, I'd say its better to let package managers do their job, and we just focus more on what we do well.

In terms of your points:

> We know to which sites the packages should go to, so not all mirrors need to have every package or every version (also we can do JIT or just ahead of time distribution)

you can configure a lot of this with within the package manager. It wont be as seamless, but I'd posit that its close enough.

> We can apply retention policies based on releases

you can do that in a package manager, but it would not be hooked up and as nice as if it was tied together
 
> We can do delta compression

So can package managers

> We can restrict those feeds so that packages from project A can't be seen by project B

Package managers can do this too, just with a different model around it.

I'd suggest that this is an optimisation rather than a core requirement of ODCM. For now, I'd just outsource the problem to the tools that specialise in it, and we can revisit later if need be.

**Comment by droyad** [comment]
I wonder why some of our customers are still having problems with this and asking us to solve it. Perhaps it is just a lack of documentation.

**Comment by matt-richardson** [comment]
Is it a case of they just don't know these things exist? Definitely interested to understand their reasoning.

**Comment by michaelnoonan** [comment]
@matt-richardson it sounds like you've done this before? It would be interesting to see a real-life implementation, poke around at it a bit and see where the friction actually is. If using a 3rd party package manager can be made reasonably seamless in an Octopus world, we could document the pattern + guide for our customers today.

In other words, I don't want to see us bet the bank on a 3rd party package manager without giving it a red-hot go first. Once we "get into bed" with an external solution we will end up beholden to that decision, and any decisions the 3rd party makes based on their own needs. If we build/maintain our own implementation, we can iterate on it as we see fit.

**Comment by michaelnoonan** [comment]
@droyad:
> Instead of having an aggregator, why not build that right into the tentacle? 
I've only just started thinking about this... but one reason why an aggregator could help is lower latency for orchestrating deployments, not just shipping logs. If your OD Server is in AU but the DC for the deployment is in UK, everything OD orchestrates has to be shipped back and forth. Placing either a Space (think Remote Release Promotions), or some kind of edge orchestrator in the DC could be a performance win.

@zentron: It sounds very similar to what Microsoft have done with IoT Edge. I could imagine a scenario one day where we split the "web server" side of Octopus Server apart from the "deployment orchestration" side. Perhaps then we could package the "deployment orchestration" engine up as `Octopus Edge`, and give it the smarts to background-replicate logs to the originating Space?

I also feel like we could achieve a reasonable approximation of this using the full Octopus Server as it is today, but by having ODCM manage the installation of the Space into the remote data center, and use the fully connected mode of Remote Release Promotions.

I really like the idea of what you've suggested. In the meantime I'd like to experiment with how ODCM + Spaces + Release Promotions plays out - if the friction is sufficiently low it might turn out to be a winner with no extra work.

**Comment by zentron** [comment]
@michaelnoonan The seperation to me sounds like the natural "solution" for geo-distribution (not to mention solving future requirements).
 Spaces might provide some help in this endeavor, but the conceptual and actual disjointness\friction that it brings from how I would expect it to work as a user feels like its more of a case of "it just happens to make things a little better for" rather than "it is made to solve the problem of". I think its just marketing of it as the solution for this particular problem that made me speak up. 

That being said the fact that it "makes things a little better for" is a good enough reason to focus our attention on the current approach as being good enough for now. :)

**Comment by thebigmoosey** [comment]
Support ticket related to this very sort of thing: https://secure.helpscout.net/conversation/439986944?folderId=557077
